rotors_gazebo_plugins/src/gazebo_wind_plugin.cpp

```diff
@@ -99,38 +113,115 @@ void GazeboWindPlugin::OnUpdate(const common::UpdateInfo& _info) {
+    wrench_msg.wrench.force.z = wind.z + wind_gust.z;
+    wrench_msg.wrench.torque.x = 0;
+    wrench_msg.wrench.torque.y = 0;
+    wrench_msg.wrench.torque.z = 0;
+
+    wind_pub_.publish(wrench_msg);
+
+    // Calculate the wind speed
+    double wind_speed = wind_speed_mean_; // + wind_speed_n_(random_generator_);
+    wind_velocity = wind_speed * wind_direction_;
+  } else {
+    // Get the current position of the aircraft in world coordinates
+    math::Vector3 link_position = link_->GetWorldPose().pos;
+
+    // Find the x and y indices of the eight vertices of the cell enclosing the aircraft
+    int i_inf = floor((link_position.x - min_x_) / res_x_);
+    int i_sup = i_inf + 1;
+    int j_inf = floor((link_position.y - min_y_) / res_y_);
+    int j_sup = j_inf + 1;
+
+    int idx_i[8], idx_j[8];
+    idx_i[0] = idx_i[1] = idx_i[4] = idx_i[5] = i_inf;
+    idx_i[2] = idx_i[3] = idx_i[6] = idx_i[7] = i_sup;
+    std::fill_n(&idx_j[0], 4, j_inf);
+    std::fill_n(&idx_j[4], 4, j_sup);
+
+    // Find the vertical factor of the aircraft in each of the four surrounding grid columns, and their minimal/maximal value
+    float vertical_factors_columns[4];
+    for (int i = 0; i < 4; i++) {
+      vertical_factors_columns[i] = (link_position.z - bottom_z_[idx_i[2*i] + (idx_j[2*i] * n_x_)]) / (top_z_[idx_i[2*i] + (idx_j[2*i] * n_x_)] - bottom_z_[idx_i[2*i] + (idx_j[2*i] * n_x_)]);
+    }
 
-  geometry_msgs::WrenchStamped wrench_msg;
+    float vertical_factors_min = std::min(std::min(std::min(vertical_factors_columns[0],vertical_factors_columns[1]),vertical_factors_columns[2]),vertical_factors_columns[3]);
+    float vertical_factors_max = std::max(std::max(std::max(vertical_factors_columns[0],vertical_factors_columns[1]),vertical_factors_columns[2]),vertical_factors_columns[3]);
 
-  wrench_msg.header.frame_id = frame_id_;
-  wrench_msg.header.stamp.sec = now.sec;
-  wrench_msg.header.stamp.nsec = now.nsec;
-  wrench_msg.wrench.force.x = wind.x + wind_gust.x;
-  wrench_msg.wrench.force.y = wind.y + wind_gust.y;
-  wrench_msg.wrench.force.z = wind.z + wind_gust.z;
-  wrench_msg.wrench.torque.x = 0;
-  wrench_msg.wrench.torque.y = 0;
-  wrench_msg.wrench.torque.z = 0;
+    // Check if aircraft is out of wind field or not, and act accordingly
+    if (!( i_inf < 0 || j_inf < 0 || vertical_factors_max < 0 || i_sup > (n_x_ - 1) || j_sup > (n_y_ - 1) || vertical_factors_min > 1 )) {
+      // Find indices in z-direction for each of the vertices. If link is not within the range of one of the columns, set indices either to lowest of highest two.
+      int idx_k[8] = {0};
+      idx_k[1] = idx_k[3] = idx_k[5] = idx_k[7] = vertical_spacing_factors_.size() - 1;
 
-  wind_pub_.publish(wrench_msg);
+      for (int i = 0; i < 4; i++) {
+        if (vertical_factors_columns[i] < 0) {
+          idx_k[2*i+1] = 1;
+        } else if (vertical_factors_columns[i] > 1) {
+          idx_k[2*i] = vertical_spacing_factors_.size() - 2;
```

Comment:
Add comment à la "Point above top grid point at this vertical column" (or similar)